rest/diagnostic_api.go

```diff
@@ -0,0 +1,341 @@
+//  Copyright 2013-Present Couchbase, Inc.
+//
+//  Use of this software is governed by the Business Source License included
+//  in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+//  in that file, in accordance with the Business Source License, use of this
+//  software will be governed by the Apache License, Version 2.0, included in
+//  the file licenses/APL2.txt.
+
+package rest
+
+import (
+	"github.com/couchbase/sync_gateway/auth"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbase/sync_gateway/channels"
+	"github.com/gorilla/mux"
+)
+
+// each field in this response is a map of collection names to a map of channel names to grant history info
+type getAllChannelsResponse struct {
+	AdminGrants       map[string]map[string]auth.GrantHistory            `json:"admin_grants,omitempty"`
+	DynamicGrants     map[string]map[string]auth.GrantHistory            `json:"dynamic_grants,omitempty"`
+	JWTGrants         map[string]map[string]auth.GrantHistory            `json:"jwt_grants,omitempty"`
+	AdminRoleGrants   map[string]map[string]map[string]auth.GrantHistory `json:"admin_role_grants,omitempty"`
+	DynamicRoleGrants map[string]map[string]map[string]auth.GrantHistory `json:"dynamic_role_grants,omitempty"`
+}
+
+func (h *handler) handleGetAllChannels() error {
+	h.assertAdminOnly()
+	user, err := h.db.Authenticator(h.ctx()).GetUser(internalUserName(mux.Vars(h.rq)["name"]))
+	if err != nil {
+		return err
+	}
+	if user == nil {
+		return kNotFoundError
+	}
```

Comment:
I think its more consistent to return 404 since /_user/_session returns 404 for an invalid user. There is a test case for that in https://github.com/couchbase/sync_gateway/blob/48c78b8fc2e7f9c16150f7514dab887b4d8f037c/rest/diagnostic_api_test.go#L249-L252